### PR TITLE
Fix for Apple Silicon

### PR DESCRIPTION
**Pull request by hayesgm**
This patch updates rust-rocksdb to correctly build on Apple Silicon (using all native tools). We include patches to rustdb found here: https://github.com/facebook/rocksdb/pull/7714. Additionally, we upgrade the build script and rocksdb libraries to fix compilation. See issue here for more details on the fixes: https://github.com/rust-rocksdb/rust-rocksdb/issues/482

Note: this is currently using a patched version of rustdb, we should wait until that PR is merged into rocksdb master until considering merging this patch.

### CODE REVIEW

File: .gitmodules
Comment (aleksuss) [inline .gitmodules:6]: I think it's a bad idea to use a fork rather than official repository.

Comment (hayesgm) [reply .gitmodules:6]: Definitely. This is meant as a placeholder until that patch is merged into mainline in rocksdb.

Comment (jasl) [reply .gitmodules:6]: https://github.com/facebook/rocksdb/commit/ee4bd4780b321ddb5f92a0f4eb956f2a2ebd60dc

Comment (hayesgm) [reply .gitmodules]: Pushed a commit to use proper FB version again

**Comment by jasl** [comment]
BTW, my friend give me another fork: <https://github.com/hdevalence/rust-rocksdb/commit/33cd4281570b229b628d6d827fa0053cff57000a>

I saw he just change `build.rs`, but this PR also change `librocksdb-sys/rocksdb_lib_sources.txt`

**Review by stanislav-tkach — APPROVED** [review]

**Comment by lerencao** [comment]
https://github.com/facebook/rocksdb/pull/7714 is already merged, so when will this happen? :)

**Comment by ordian** [comment]
if you're updating the rocksdb submodule, don't forget to update the `build_version.cc`:
```patch
diff --git a/librocksdb-sys/build_version.cc b/librocksdb-sys/build_version.cc
index 03e7541..f4328c8 100644
--- a/librocksdb-sys/build_version.cc
+++ b/librocksdb-sys/build_version.cc
@@ -1,4 +1,4 @@
 #include "build_version.h"
-const char* rocksdb_build_git_sha = "rocksdb_build_git_sha:@48bfca38f6f175435052a59791922a1a453d9609@";
-const char* rocksdb_build_git_date = "rocksdb_build_git_date:@2020/07/15 17:54:15@";
+const char* rocksdb_build_git_sha = "rocksdb_build_git_sha:@eee0af9af179063f189680e2bd79900de2f9fc6f@";
+const char* rocksdb_build_git_date = "rocksdb_build_git_date:@2020/12/05 14:17:11@";
 const char* rocksdb_build_compile_date = __DATE__;
 ```
 I assume we're waiting for a rocksdb release containing this fix?

**Comment by niklaslong** [comment]
FYI there's a [new rocksdb release](https://github.com/facebook/rocksdb/releases/tag/v6.15.2) containing the fix.

**Comment by Kixunil** [comment]
Did anyone happen to benchmark this? Nothing serious, I'm just curious. :)

**Comment by aleksuss** [comment]
There is a `rocksdb 6.15.4` in the master branch. @hayesgm try to update the branch.

**Comment by jasl** [comment]
> There is a `rocksdb 6.15.4` in the master branch. @hayesgm try to update the branch.

I'm confusing about RocksDB releases, Apple Silicon are not included in 6.15.x,
according to https://github.com/facebook/rocksdb/blob/master/HISTORY.md ,
6.16.0 and 6.17.0 already versioned, but they didn't list to GitHub Releases page

**Comment by ordian** [comment]
A new rocksdb release is available: https://github.com/facebook/rocksdb/releases

Can we please merge this soon?

**Comment by ordian** [comment]
sorry, couldn't help myself #503 :smile:

**Comment by hayesgm** [comment]
Awesome; thanks for following up @ordian. Closing in lieu of #503.

**Comment by jvsteiner** [comment]
I'm still getting this on a much newer version:

```
error: failed to run custom build command for `librocksdb-sys v0.8.0+7.4.4`

Caused by:
  process didn't exit successfully: `/Users/jamie/Code/substrate-node-template/target/release/build/librocksdb-sys-5135db74e4151ca7/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "the `libclang` shared library at /opt/homebrew/Cellar/llvm/15.0.6/lib/libclang.dylib could not be opened: dlopen(/opt/homebrew/Cellar/llvm/15.0.6/lib/libclang.dylib, 0x0005): tried: '/usr/local/Cellar/mysql@5.6/5.6.35/lib/libclang.dylib' (no such file), '/libclang.dylib' (no such file), '/opt/homebrew/Cellar/llvm/15.0.6/lib/libclang.dylib' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))"', /Users/jamie/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.60.1/src/lib.rs:2172:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This is a close issue, but still seems to be happening?

**Pull request closed without merging**
